pilot/pkg/proxy/envoy/v2/rds.go

```diff
@@ -53,10 +55,24 @@ func (s *DiscoveryServer) pushRoute(con *XdsConnection, push *model.PushContext,
 
 func (s *DiscoveryServer) generateRawRoutes(con *XdsConnection, push *model.PushContext) []*xdsapi.RouteConfiguration {
 	rc := make([]*xdsapi.RouteConfiguration, 0)
+	// collect workload labels
+	var workloadLabels labels.Collection
+	for _, w := range con.modelNode.ServiceInstances {
+		workloadLabels = append(workloadLabels, w.Labels)
+	}
+
+	gateways := s.Env.Gateways(workloadLabels)
+	if len(gateways) == 0 {
+		log.Debuga("buildGatewayRoutes: no gateways for router ", con.modelNode.ID)
+		return nil
+	}
+
+	merged := model.MergeGateways(gateways...)
+	log.Debugf("buildGatewayRoutes: gateways after merging: %v", merged)
```

Comment:
We need this only for router type proxy, right?

Reply:
@hzxuzhonghu yes, sidecar mode is not required hence I did not change any on that path.

Reply:
As this is a common part for both `sidecar` and `router` type proxies, should make this only run for `router`

Reply:
the loop is shared, and the logic need to be out of the loop. that's why it is passed to the common function and only applied to router path.
otherwise, different type need different loop.

Reply:
But we need to make sure no extra cal on `sidecar` proxies, just skip these above for them.

Reply:
Several things: you should not need to do the workload labels thing. Use node.workload labels 

Second, see if you can cache the list of all gateways in the push context , just like we cache things for Sidecar, EnvoyFilter, etc. then you can simply do push.Gateways(node) rather than env.Gateways which is a costly operation (I don’t know how we missed this!!)

Third, if the merge gateways is a costly operation, add a simple push.InitNodeConfigs(node)
In this function, you can check node type, compute the merged gateways, and store them in the node itself just like the service instances etc.

Reply:
And yes, you need an if block in the code to do this computation only for routers